[Docs] Example of using Vizro-AI in chain as tools (#841)
mckinsey/vizro: vizro-ai/docs/pages/user-guides/vizro-ai-langchain-guide.md

# Using Vizro-AI methods as LangChain tools

> This guide demonstrates how to integrate Vizro-AI's chart and dashboard generation capabilities as LangChain tools. This integration enables you to use Vizro-AI's functionality within a larger LangChain application.

Reviewer: ```suggestion
You can use Vizro-AI's functionality within a larger LangChain application. This guide shows how to integrate Vizro-AI's chart and dashboard generation capabilities as LangChain tools 
```

Reviewer: This is great. I would be tempted to add another sentence or two to say, these are the steps that are necessary, and then include a number-bulleted list of each of them, using the subsection headings, and linking to the section of the page.

So, something like the following suggestion:

Reviewer: ```suggestion
These are the steps you need to take:

1. [Set up the environment](#whatever-this-is)
2. [Define LangChain tools](#whatever-this is)
3. etc
```

## 1. Set up the environment

First, import the required libraries and prepare your data:

```python
from copy import deepcopy
from typing import Annotated, Any

import pandas as pd
import vizro.plotly.express as px
from langchain_core.runnables import chain
from langchain_core.tools import InjectedToolArg, tool
from langchain_openai import ChatOpenAI
from vizro_ai import VizroAI

# Load sample data
df = px.data.gapminder()
dfs = [df]

# Initialize the LLM
llm = ChatOpenAI(model="gpt-4o-mini")
```

## 2. Define LangChain tools

> Basic tools only take string as input and output. Vizro-AI takes Pandas dataframes as input and it's not cost-efficient and secure to pass the full data to a LLM. In this case, [bind the Pandas dataframes at run time](https://python.langchain.com/v0.2/docs/how_to/tool_runtime/) is more suitable.

Reviewer 2: ```suggestion
Basic tools only take string as input and output. Vizro-AI takes Pandas DataFrames as input and it's neither cost-efficient nor secure to pass the full data to a LLM. In this case, [binding the Pandas dataframes at run time](https://python.langchain.com/v0.2/docs/how_to/tool_runtime/) is more suitable.
```

> Basic tools only take string as input and output. Vizro-AI takes Pandas dataframes as input and it's not cost-efficient and secure to pass the full data to a LLM. In this case, [bind the Pandas dataframes at run time](https://python.langchain.com/v0.2/docs/how_to/tool_runtime/) is more suitable.

Reviewer: ```suggestion
Vizro-AI takes Pandas dataframes as input, but it's not cost-efficient and secure to pass the full data to a LLM since basic tools only take strings as input and output. In this step, [bind the Pandas dataframes at run time](https://python.langchain.com/v0.2/docs/how_to/tool_runtime/).
```

Author: Current steps 2 and 3 together make the approach complete. Now I reword the sentence as "The recommended approach is to exclude DataFrame parameters from the tool's schema (which is the step2) and instead bind them at runtime (which is the step3) using LangChain's runtime binding feature". Would this be more clear?

    """Generate only the plot code.

    Args:
        df: A pandas DataFrame
        question: The plotting question

    Returns:
        Generated plot code
    """
    df = pd.DataFrame(df)
    vizro_ai = VizroAI(model=llm)
    plot_elements = vizro_ai.plot(
        df,
        user_input=question,
        return_elements=True,
    )
    return plot_elements.code_vizro

@tool(parse_docstring=True)
def get_dashboard_code(dfs: Annotated[Any, InjectedToolArg], question: str) -> str:
    """Generate the dashboard code.

    Args:
        dfs: Pandas DataFrames
        question: The dashboard question

    Returns:
        Generated dashboard code
    """
    vizro_ai = VizroAI(model=llm)
    dashboard_elements = vizro_ai.dashboard(
        dfs,
        user_input=question,
        return_elements=True,
    )
    return dashboard_elements.code
```

## 3. Set up the tool chain

Create a chain that handles tool execution and data injection:

```python
# Bind tools to the LLM
tools = [get_plot_code, get_dashboard_code]
llm_with_tools = llm.bind_tools(tools)

# Create data injection chain
@chain
def inject_df(ai_msg):
    tool_calls = []
    for tool_call in ai_msg.tool_calls:

>         tool_call_copy = deepcopy(tool_call)

Reviewer 2: Why is the deepcopy needed?

Author: Didn't put much thoughts on it as it's borrowed from langchain docs. In general it's good practice not to touch the original `tool_calls` though

        if tool_call_copy["name"] == "get_dashboard_code":
            tool_call_copy["args"]["dfs"] = dfs
        else:
            tool_call_copy["args"]["df"] = df

        tool_calls.append(tool_call_copy)
    return tool_calls

# Create tool router
tool_map = {tool.name: tool for tool in tools}

@chain
def tool_router(tool_call):
    return tool_map[tool_call["name"]]

# Combine chains
chain = llm_with_tools | inject_df | tool_router.map()
```

## 4. Use the chain

Now you can use the chain to generate charts or dashboards based on natural language queries. The chain will generate code that you can use to create visualizations.

!!! example "Generate chart code"

    === "Code"
        ```py

>         plot_response = chain.invoke("Plot GDP per capita for each continent")

Reviewer 2: How would I invoke the chain with a different df, can that we passed as an argument, or does this only work if hard-coded at the top?

Author: I think the key is which data `df` is pointing to at run time of `chain.invoke()`. So we can do the following:
```
df = px.data.gapminder()
plot_response = chain.invoke("Plot GDP per capita for each continent")

df = px.data.iris()
plot_response = chain.invoke("Plot the distribution of petal length")
```
